src/transformStateWithClones.js

```diff
@@ -1,13 +1,33 @@
 'use strict';
 
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    for (const key in action) {
+      if (action[key] === 'addProperties') {
+        Object.assign(newState, action.extraData);
```

Comment:
Ensure that `structuredClone` is supported in your environment, as it is a relatively new feature and may not be available in all JavaScript environments. If not supported, consider using an alternative deep cloning method.